zeroeq/subscriber.cpp

```diff
@@ -178,59 +130,31 @@ class Subscriber::Impl
     {
-        zmq_msg_t msg;
-        zmq_msg_init( &msg );
-        zmq_msg_recv( &msg, socket.socket, 0 );
-
-        uint128_t type;
-        memcpy( &type, zmq_msg_data( &msg ), sizeof(type) );
-#ifndef ZEROEQ_LITTLEENDIAN
-        detail::byteswap( type ); // convert from little endian wire
-#endif
-        const bool payload = zmq_msg_more( &msg );
-        zmq_msg_close( &msg );
-
-        SerializableMap::const_iterator i = _serializables.find( type );
-        if( i == _serializables.end( )) // FlatBuffer
-        {
-            zeroeq::Event event( type );
-            if( payload )
-            {
-                zmq_msg_init( &msg );
-                zmq_msg_recv( &msg, socket.socket, 0 );
-                const size_t size = zmq_msg_size( &msg );
-                ConstByteArray data( new uint8_t[size],
-                                     std::default_delete< uint8_t[] >( ));
-                memcpy( (void*)data.get(), zmq_msg_data( &msg ), size );
-                event.setData( data, size );
-                assert( event.getSize() == size );
-                zmq_msg_close( &msg );
-            }
-
-            if( _eventFuncs.count( type ) != 0 )
-                _eventFuncs[type]( event );
-#ifndef NDEBUG
-            else
-            {
-                // Note eile: The topic filtering in the handler registration
-                // should ensure that we don't get messages if we have no
-                // handlers. If this throws, something does not work.
-                ZEROEQTHROW( std::runtime_error( "Got unsubscribed event" ));
-            }
-#endif
-        }
-        else // serializable
-        {
-            servus::Serializable* serializable = i->second;
-            if( payload )
-            {
-                zmq_msg_init( &msg );
-                zmq_msg_recv( &msg, socket.socket, 0 );
-                serializable->fromBinary( zmq_msg_data( &msg ),
-                                          zmq_msg_size( &msg ));
-                zmq_msg_close( &msg );
-            }
-        }
+       zmq_msg_t msg;
+       zmq_msg_init( &msg );
+       zmq_msg_recv( &msg, socket.socket, 0 );
+
+       uint128_t type;
+       memcpy( &type, zmq_msg_data( &msg ), sizeof(type) );
+    #ifndef ZEROEQ_LITTLEENDIAN
+       detail::byteswap( type ); // convert from little endian wire
+    #endif
+       const bool payload = zmq_msg_more( &msg );
+       zmq_msg_close( &msg );
+
+       SerializableMap::const_iterator i = _serializables.find( type );
+       if( i == _serializables.cend( ))
+           ZEROEQTHROW( std::runtime_error( "Got unsubscribed event " + type.getString( )));
+
+       servus::Serializable* serializable = i->second;
+       if( payload )
+       {
+           zmq_msg_init( &msg );
+           zmq_msg_recv( &msg, socket.socket, 0 );
+           serializable->fromBinary( zmq_msg_data( &msg ),
+                                     zmq_msg_size( &msg ));
+           zmq_msg_close( &msg );
+       }
```

Comment:
@eile opened an issue for this #158. Will be handled after some discussions. Current fromBinary() function is implicitly calling notification. It wont be called for events without a payload.